Asynchronous prepare step & status downtime

skazi0 opened the pull request:
The "prepare" call was changed to asynchronous in the backend (i.e. step is performed in the background).
The UI was adjusted to poll for step status before switching to next step.

This PR includes some refactoring for the previous implementation of polling to make it re-usable and optimize tests.

In addition separate commit introduces option to specify allowed downtime for the status service to cover cases where backend which provides status information becomes temporarily unavailable.

skazi0 committed:
upgrade: Extracted polling for current step's status

The polling/waiting for current step to end needs to be used in
multiple pages so to avoid code duplication and simplify testing
this functionality was extracted into a separate service.

Admin server upgrade step was updated to use the new service.
Other places where it needs to be used will be done separately.

skazi0 commented:
![](https://github.trello.services/images/mini-trello-icon.png) [(5) S50P1 (S49P11) perform the prepare in the background (bsc#997293)](https://trello.com/c/YFa8WWef/144-5-s50p1-s49p11-perform-the-prepare-in-the-background-bsc-997293)

skazi0 committed:
upgrade: Updated Landing page to poll for prepare status

The prepare operation is done in background so the UI needs to
poll for status of this step to keep user on the landing page until
prepare is finished.

skazi0 commented:
Note that it's still based on https://github.com/crowbar/crowbar-ui/pull/83 so rebase will be needed before merging.

skazi0 committed:
upgrade: Allow temporary downtime when polling for status

In some cases (e.g. admin server upgrade) the status API becomes
unavailable temporarily due to node reboot or service restart.
Those downtimes should not be treated as errors if they don't last
too long.

This change introduces additional parameter to waitForStepToEnd()
function to enable caller to specify how much downtime is acceptable.
First usage for this is admin server upgrade step mentioned above.

skazi0 commented:
![](https://github.trello.services/images/mini-trello-icon.png) [(3) S50P10 (S49P15) #11 Upgrade Admin Server: UI Story 2 - Service Integration](https://trello.com/c/PAkqYMEe/147-3-s50p10-s49p15-11-upgrade-admin-server-ui-story-2-service-integration)

skazi0 committed:
upgrade: Fix ongoing upgrade status handling

When page was reloaded during admin server upgrade the upgrade
button would get disabled and start spinning but polling for status
would not be (re)started.

This change extracts the common "wait for upgrade step to end" part
into a separate function and uses it in both cases (new and ongoing
upgrade).

santiph commented on assets/app/data/crowbar/services/upgrade-status.factory.js:
```diff
@@ -0,0 +1,69 @@
+        }
+
+        /**
+         *
```
missing docblock

santiph commented on assets/app/data/crowbar/services/upgrade-status.factory.js:
```diff
@@ -0,0 +1,69 @@
+        function getStatus() {
+            var requestOptions = {
+                method: 'GET',
+                url: '/api/upgrade',
```
this is part of the upgrade factory. Please move it there

santiph commented on assets/app/data/crowbar/services/upgrade-status.factory.spec.js:
```diff
@@ -0,0 +1,235 @@
+        describe('waitForStepToEnd function', function () {
+            it('should be defined', function () {
+                should.exist(upgradeStatusFactory.waitForStepToEnd);
+                expect(upgradeStatusFactory.waitForStepToEnd).toEqual(jasmine.any(Function));
```
Same as line 176

skazi0 replied:
removed the other one.

santiph commented on assets/app/features/upgrade/controllers/upgrade-landing.controller.js:
```diff
@@ -82,7 +97,20 @@
-                    $state.go('upgrade.backup');
+                    vm.prepare.running = true;
+                    upgradeStatusFactory.waitForStepToEnd(
+                        'upgrade_prepare',
```
We should have a centralised map to store the steps values and reuse them anywhere in the flow

santiph reviewed (changes requested):
Partially reviewed. I'm still missing some files

skazi0 committed:
upgrade: moved string literals to constants

Some strings like step names and statuses are used in multiple places
so it is better to have them defined as constants. With this approach
the compiler will help spot possible typos which will be harder to spot
with literal string.

Itxaka commented on assets/app/data/crowbar/services/upgrade-status.factory.js:
```diff
@@ -0,0 +1,60 @@
+            upgradeFactory.getStatus()
+                .then(
+                    function (response) {
+                        if (response.data.steps[step].status == 'passed') {
```
cant we use `STEP_STATES.passed` in the comparision?

skazi0 replied:
This is a bit tricky, `STEP_STATES` are defined in `upgrade.constants.js` which is loaded after the data services. I didn't find any better place to put these constants which would make them accessible to the `upgradeStatusFactory`. Any idea for this?

Itxaka replied:
Mmmh, I dont think that we should change it then. Doing acrobatics with the code just to use the constant in one place seems too much.

Itxaka commented:
Wow! looking pretty good to me, only a small question but looking great!

santiph commented on assets/app/data/crowbar/services/upgrade-status.factory.js:
```diff
@@ -0,0 +1,60 @@
+        /**
+         * Polls for upgrade status until step `step` is `passed`.
+         *
+         * Polling is done at given `pollingInterval` interval.
```
Great job explaining parameters!
Please, use the following to write docblocks: https://github.com/angular/angular.js/wiki/Writing-AngularJS-Documentation
something like:
```
@param {type} step description
@param {type} onSuccess callback to be executed with last response from status API when waiting time finishes successfully.
@param {type} onError description
@param {number} pollingInterval Interval used to poll the upgrade status
@param {type} allowedDowntimeLeft description
```

santiph commented on assets/app/features/upgrade/controllers/upgrade-upgrade-administration-server.controller.js:
```diff
@@ -61,34 +78,24 @@
+        function waitForUpgradeToEnd() {
+            upgradeStatusFactory.waitForStepToEnd(
+                UPGRADE_STEPS.admin_upgrade,
+                function (/*response*/) {
```
Could you add a comment clarifying this will be passed as success handler to the last promise?

skazi0 replied:
Do you write a comment for each callback you pass to standard `when()`? Client code should not explain how library works. The library's docs are for that.

santiph replied:
Actually, I do because I think it helps new angular developers to understand and read the code more clearly. On my experience, understanding Promises is one of the most frequent challenges people get...

But I see your point. It was definitely not a blocker, but a suggestion.

santiph commented on assets/app/features/upgrade/upgrade.constants.js:
```diff
@@ -12,5 +12,26 @@
         })
-        .constant('ADMIN_UPGRADE_TIMEOUT_INTERVAL', 1000);
+        .constant('PREPARE_TIMEOUT_INTERVAL', 1000)
+        .constant('ADMIN_UPGRADE_ALLOWED_DOWNTIME', 5 * 60 * 1000)
```
Where did you get this number from?

skazi0 replied:
5 minutes is just a blind guess for the time needed to reboot the admin server. This needs to be adjusted during testing.

skazi0 replied:
@vuntz suggested to change these to poll every 5-10sec and tolerate 20-30min downtime. PR updated.

santiph reviewed (changes requested):
minor clarifications requested

skazi0 commented:
rebased.

santiph commented:
I've sent an email requesting clarification on the timeout requirements.
Once confirmed, consider this approved.
Thanks for the hard work!

skazi0 commented:
![](https://github.trello.services/images/mini-trello-icon.png) [(2) S51P9 #Crowbar-ui: re-integrate landing page with latest API changes](https://trello.com/c/syDbFm5C/186-2-s51p9-crowbar-ui-re-integrate-landing-page-with-latest-api-changes)
